syntax = "proto3";
//option optimize_for = LITE_RUNTIME;

import "MaterialScriptObject.proto";
import "ModelScriptObject.proto";
import "SceneScriptObject.proto";
import "MetaScriptObject.proto";


package Tiny3D.Script.FileFormat;


message FileHeader
{
    enum FileType
    {
        Material = 0;
        Model = 1;
        Scene = 2;
        Meta = 3;
    }
    
    string      magic = 1;
    FileType    type = 2;
    fixed32     version = 3;
}


message FileMaterial
{
    FileHeader              header = 1;
    MaterialSystem.Material material = 2;
}


message FileMeta
{
    FileHeader          header = 1;
    MetaSystem.MetaData meta = 2;
}

message FileModelData
{
    string                                  root = 1;
    map<string, SceneSystem.Node>           nodes = 2;
    map<string, ModelSystem.MeshData>       meshes = 3;
    map<string, ModelSystem.AnimationClip>  animations = 4;
}

message FileModel
{
    FileHeader      header = 1;
    FileModelData   data = 2;
}

message FileSceneData
{
    string                              root = 2;
    map<string, SceneSystem.Node>       nodes = 3;
    map<string, ModelSystem.MeshData>   meshes = 4; 
}

message FileScene
{
    FileHeader          header = 1;
    FileSceneData       data = 2;
}
